### PR TITLE
feat(exex, primitives): serde bincode compatibility

### LINKED ISSUE #11170: Bincode support for `ExExNotification`
### Describe the feature

Blocked by https://github.com/alloy-rs/alloy/issues/1349

## Problem

[Alloy](https://github.com/alloy-rs/alloy) has a bunch of structs that use `#[serde(skip*)]` attributes. This attributes isn't supported in [bincode](https://github.com/bincode-org/bincode/issues/326), as well as in [rmp-serde](https://github.com/3Hren/msgpack-rust/issues/86).

Unfortunately, It means that we cannot serialize/deserialize `ExExNotification` using these formats.

## Solution

Introduce helper serde modules for those structs that are downstream from `ExExNotification` and have `#[serde(skip*)]` attributes on fields and remove those attributes. Modules should be similar to https://github.com/alloy-rs/alloy/blob/5c7470c113df23a625456413bdf8c87e0b7438bc/crates/rpc-types-beacon/src/payload.rs#L275-L298

Then, these modules should be used in structs that have problematic types as fields. For example, for the `Block` type we will need to use a `#[serde_as]` on the `Block` struct and `#[serde_as(as = "HeaderBincode")]` on the `header` field https://github.com/paradigmxyz/reth/blob/4da5f1f388d8a1ebf112e094bf272566238efc24/crates/primitives/src/block.rs#L34-L37

Both of the attributes should be [gated](https://docs.rs/serde_with/latest/x86_64-pc-windows-msvc/serde_with/guide/serde_as/index.html#gating-serde_as-on-features) by the `serde-bincode` feature.

The types that need special handling are:

- `ExExNotification`
- `Chain`
- `SealedBlockWithSenders`
- `SealedBlock`
- `SealedHeader`
- `BlockBody`
- `TransactionSigned`
- `Transaction`

As a test, we should be able to serialize and deserialize the `ExExNotification` struct using `bincode`.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shekhirin**
Closes https://github.com/paradigmxyz/reth/issues/11170

Blocked by https://github.com/alloy-rs/op-alloy/pull/131

### CODE REVIEW

**Comment by shekhirin** [comment]
fails with optimism feature, fixing

```console
❯ SEED=4 cargo t --quiet -p reth-primitives --features serde-bincode-compat,optimism block_body_bincode

running 1 test
block::serde_bincode_compat::tests::test_block_body_bincode_roundtrip --- FAILED

failures:

---- block::serde_bincode_compat::tests::test_block_body_bincode_roundtrip stdout ----
thread 'block::serde_bincode_compat::tests::test_block_body_bincode_roundtrip' panicked at crates/primitives/src/block.rs:1086:64:
called `Result::unwrap()` on an `Err` value: InvalidTagEncoding(32)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    block::serde_bincode_compat::tests::test_block_body_bincode_roundtrip

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 99 filtered out; finished in 0.09s

error: test failed, to rerun pass `-p reth-primitives --lib`

❯ SEED=4 cargo t --quiet -p reth-primitives --features serde-bincode-compat block_body_bincode 

running 1 test
.
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 96 filtered out; finished in 0.10s
```

**Comment by shekhirin** [comment]
fixed by https://github.com/alloy-rs/op-alloy/pull/131

```console
❯ cargo t -p reth-primitives --features serde-bincode-compat,optimism --no-run                  
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.34s
  Executable unittests src/lib.rs (target/debug/deps/reth_primitives-5d9f9518ba4cbf81)

❯ hyperfine -r 100 -- "./target/debug/deps/reth_primitives-5d9f9518ba4cbf81 block_body_bincode"
Benchmark 1: ./target/debug/deps/reth_primitives-5d9f9518ba4cbf81 block_body_bincode
  Time (mean ± σ):      63.6 ms ±   6.9 ms    [User: 59.5 ms, System: 1.6 ms]
  Range (min … max):    59.1 ms … 113.2 ms    100 runs
```

File: crates/evm/execution-types/src/chain.rs [hunk 1]
```diff
@@ -723,3 +723,124 @@ mod tests {
...
+    }
+
+    impl<'a> From<&'a super::Chain> for Chain<'a> {
+        fn from(value: &'a super::Chain) -> Self {
+            Self {
+                blocks: value
+                    .blocks
+                    .iter()
+                    .map(|(block_number, block)| (*block_number, block.into()))
+                    .collect(),
```
Comment (mattsse) [inline crates/evm/execution-types/src/chain.rs:770]: we could get rid of this by manually impl serialize but for now this is okay

**Review by mattsse — APPROVED** [review]
this isn't super pretty but should do

**Pull request merged**